Avoid hanging at exit when using CLONE_VM on Glibc <2.24

blue42u opened the pull request:
Fixes #455, ping @jmellorcrummey and @mwkrentel. IMO 2.24 is old enough we could get away with just dropping support, if not though this PR provides a hacky fix.

There is a bug in older Glibc (https://sourceware.org/bugzilla/show_bug.cgi?id=18862) where using `CLONE_VM` instead of `CLONE_THREAD` causes the Glibc wrapper to reset its internal pthread state in the parent process. This wreaks havoc with all things pthread, but most noticably it stops the `pthread_kill` in libmonitor's shootdown from functioning (and libmonitor doesn't pay attention enough to the returned error codes).

As a sort of hack, it turns out the reset happens after the body passed to `clone`, so killing the child via a signal instead of returning dodges the issue. At least, it seems to work on 2.17.

This also adds an auditor-export for `sigaction` (to reset the signal disposition for a faster signal-death). In the fake-auditor case there isn't a "default" function to use since `sigaction` is wrapped by libmonitor, which in turn doesn't provide a `monitor_real_sigaction`. Caveat emptor.

jmellorcrummey commented on src/tool/hpcrun/fnbounds/fnbounds_client.c:
```diff
@@ -455,7 +455,17 @@ hpcfnbounds_child(void* fds_vp) {
+  act.sa_flags = 0;
+  auditor_exports->sigaction(SIGUSR1, &act, NULL);
+
+  raise(SIGUSR1);
```
This idiom doesn't work on NOTS. The raise returns and then abort terminates with a SEGV. This causes everything to die, including the application. 
- I don't know why the process that raised an exception on itself doesn't die.
- I don't know why the child's death takes down the parent.

blue42u commented:
Simple reproducer for the GDB bug:
```console
$ cat x.c
#define _GNU_SOURCE
#include <sched.h>
#include <sys/wait.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

int body(void* _) {
  struct timespec ts = {2, 0};
  while(nanosleep(&ts, &ts));
  return 0;
}

int main() {
  void* stack = malloc(16*2*1024);
  int child = clone(body, stack+(16*1024), SIGCHLD | CLONE_VM, NULL);
  printf("Waiting for child: %d\n", child);
  waitpid(child, NULL, __WALL);
  printf("Child waited for\n");
  return 0;
}
$ gcc x.c -g
$ gdb ./a.out
...
(gdb) b x.c:17
Breakpoint 1 at 0x11da: file x.c, line 17.
(gdb) b x.c:19
Breakpoint 2 at 0x1221: file x.c, line 19.
(gdb) r
Starting program: /tmp/a.out 

Breakpoint 1, main () at x.c:17
17	  int child = clone(body, stack+(16*1024), SIGCHLD | CLONE_VM, NULL);
(gdb) c
Continuing.
[Detaching after fork from child process 298576]
Waiting for child: 298576
Child waited for
[Inferior 1 (process 298566) exited normally]
(gdb)
```

Weirdly enough forcing GDB to step slowly through the `clone` syscall makes it work again:
```console
(gdb) catch syscall clone
Catchpoint 3 (syscall 'clone' [56])
(gdb) r
Starting program: /tmp/a.out 

Breakpoint 1, main () at x.c:17
17	  int child = clone(body, stack+(16*1024), SIGCHLD | CLONE_VM, NULL);
(gdb) c
Continuing.

Catchpoint 3 (call to syscall clone), clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:78
78	../sysdeps/unix/sysv/linux/x86_64/clone.S: No such file or directory.
(gdb) c
Continuing.
[Detaching after fork from child process 298916]

Catchpoint 3 (returned from syscall clone), clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:78
78	in ../sysdeps/unix/sysv/linux/x86_64/clone.S
(gdb) c
Continuing.
Waiting for child: 298916

Breakpoint 2, main () at x.c:19
19	  waitpid(child, NULL, __WALL);
(gdb) c
Continuing.
Child waited for
[Inferior 1 (process 298897) exited normally]
(gdb) 
```

blue42u commented:
Updated with the "backup plan," new patch description below. Rebased on https://github.com/HPCToolkit/hpctoolkit/pull/457 to prevent conflicts during the merge.

The fix here is to remove `CLONE_VM`, which as a side effect removes the parent's tracking for the Fnbounds process. Although there is a "shutdown" message that could be sent to close Fnbounds, since sending the message may block indefinitely this is not used to signal a shutdown (unchanged from the original). In the worst case scenario of an Fnbounds hung between queries, the application will exit without concern even though Fnbounds has not yet (previously it would `waitpid` for it to exit). Caveat emptor.                     

Since this bug only affects the parent, the child and grandchild share a memory space. Neither use pthreads and both exit quickly (and through different paths) so it should be fine.

Note that this will re-raise the slowdowns observed with large memory spaces. Caveat emptor.

mwkrentel commented:
Note: although it may be a little sloppy to ignore the return code for pthread_kill(),
it doesn't really matter.  In either case, monitor will be unable to signal the thread
to run code in the thread.  So, you'll still necessarily get some message that means
"can't connect to some threads" and you'll still miss the fini_thread callback.

I've never liked the need for signals for this case.  My plan for monitor NG, if it ever
happens, is not to call fini_thread for this case, instead just tell hpcrun that the process
is ending and let hpcrun deal with it.  For example, some threads will write out other
threads data, but not have this strong need to do everything from the same thread.

mwkrentel commented:
This (and related clone issues) would make great candidates for regression tests.
